kotlinx-coroutines-test/common/src/TestScope.kt

```diff
@@ -251,6 +259,8 @@ internal class TestScopeImpl(context: CoroutineContext) :
     fun legacyLeave(): List<Throwable> {
         val exceptions = synchronized(lock) {
             check(entered && !finished)
+            /** After [finished] becomes `true`, it is no longer valid to have [reportException] as the callback. */
+            ExceptionCollector.removeOnExceptionCallback(lock)
```

Comment:
Callbacks are never removed for non-legacy version, consuming the memory lineary to number of tests run, please address that